docs/spec_draft.md

```diff
@@ -0,0 +1,124 @@
+# StableHLO Specification Draft
+
+## Types
+
+Following are the supported element types in StableHLO:
+
+  * **Integer types**
+
+    * Signed integer with two’s complement representation. Referred to in the
+    document as `si<N>`, where the bit-width N ∊ {4, 8, 16, 32, 64}
+    * Unsigned integer referred to in the document as `ui<N>`, where the
+    bit-width N ∊ {4, 8, 16, 32, 64}
+  * **Boolean types** referred to in the document as `pred`
+  * **Floating point types**
+    * Single precision `f32`, double precision `f64` and half precision `f16`
+    floating points complying with [IEEE 754
+    format](https://ieeexplore.ieee.org/document/8766229).
+    * Bfloat16 `bf16` floating point complying with [Brain Floating Point Format](https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus).
+    Provides the same number of exponent bits as `f32`, so that it matches its
+    dynamic range, but with greatly reduced precision.
+ * **Complex types** represents a pair of floating point types. Supported ones
+ are `c64` (represents paired `f32`) and `c128` (represents paired `f64`).
+
+StableHLO supports a Tensor type `tensor`, to model the type of a n-dimensional
+array, represented in IR as `tensor<SxE>` such that
+
+  * Shape `S` is a list of number of elements in each of the
+  dimensions and represented, in increasing order of the corresponding dimension
+  number, as an array of values of type `ui64`. A 0-value represents missing or
+  empty data in the corresponding dimension.
+  * Element type `E` is any one of the supported element types mentioned above.
+
+## Programs
+
+StableHLO programs consist of functions. Each function has arguments and results
+of supported types and a list of ops in static single-assignment (SSA) form
+which is terminated by a return op which produces the results of the function.
+StableHLO ops take operands and produce results.
+
+```
+func.func @example_func(%arg: tensor<4x16xf32>) -> tensor<4x16xf32>  {
```

Comment:
Let's use `ml_program.func` and `ml_program.return` in this example?

Reply:
done.